File: ord/src/main/java/dag/ord/source/snl/SnlSource.java
package dag.ord.source.snl;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import dag.ord.search.Result;
import dag.ord.source.JsonSource;
import dag.ord.source.Source;

public class SnlSource extends JsonSource {
    private static final String BASEURL = "https://[subdomene]snl.no/api/v1/search?query=[query]";
    private SubDomain subDomain;

    private SnlSource(String shortName, SubDomain subDomain) {
        super(shortName);
        this.subDomain = subDomain;
    }

    public static List<Source> createWordSources() {
        List<Source> sources = new ArrayList<Source>();
        sources.add(new SnlSource("SNL", SubDomain.SNL));
        sources.add(new SnlSource("NKL", SubDomain.NKL));
        sources.add(new SnlSource("NBL", SubDomain.NBL));
        sources.add(new SnlSource("SML", SubDomain.SML));
        return sources;
    }

    protected String getLookupUrl(String urlEncodedQueryWord) {
        return BASEURL.replace("[subdomene]", subDomain.getSubDomain()).replace("[query]",urlEncodedQueryWord);
    }

    protected List<Result> toResults(String queryWord, String json, int maxWordDescriptorLength) {
        List<Result> Results = new ArrayList<Result>();

        try {
            JSONArray articles = new JSONArray(json);

            for (int i = 0; i < articles.length(); i++) {
                JSONObject article = articles.getJSONObject(i);

                String displayUrl = toTrimmedString(article, "article_url");
                String clarification = toTrimmedString(article, "clarification");
                String headWord = toTrimmedString(article, "headword");
                String snippet = toTrimmedString(article, "snippet");
                String taxonomyTitle = toTrimmedString(article, "taxonomy_title");
                String summary = "<b>" +headWord + "</b> <i>(";
                if (clarification != null) {
                    summary += clarification + "/";
                }
                summary += taxonomyTitle + ")</i> " + snippet;

                Result result = new Result(displayUrl, summary, maxWordDescriptorLength);
                Results.add(result);
            }
        } catch (JSONException e) {
            Result result = new Result(null, e.toString(), maxWordDescriptorLength);
            Results.add(result);
        }

        return Results;
    }


    public enum SubDomain {
        SNL(""),
        NKL("nkl."),
        NBL("nbl."),
        SML("sml.");
        private String subDomain;

        SubDomain(String subDomain) {
            this.subDomain = subDomain;
        }

        String getSubDomain() {
            return subDomain;
        }
    }

}

File: ord/src/main/java/dag/ord/source/snl/SnlSource.kt
package dag.ord.source.snl

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.core.type.TypeReference
import dag.ord.search.Result
import dag.ord.source.Source
import dag.ord.util.JsonMapper
import java.util.*

private const val BASEURL = "https://[subdomene]snl.no/api/v1/search?query=[query]"

class SnlSource private constructor(sourceId: String, private val subDomain: String) : Source(sourceId) {
    override fun getLookupUrl(urlEncodedQueryWord: String): String {
        return BASEURL.replace("[subdomene]", subDomain).replace("[query]", urlEncodedQueryWord)
    }

    override fun toResults(queryWord: String, urlContent: String, maxResultLength: Int) =
            JsonMapper()
                    .read(urlContent, object : TypeReference<List<Article>>() {})
                    .map {
                        val summary = "<b>${it.headword}</b> <i>(${it.taxonomyTitle})</i> ${it.snippet}"
                        Result(it.articleUrl, summary, maxResultLength)
                    }

    companion object {
        fun createWordSources() = listOf(
            SnlSource("SNL", ""),
            SnlSource("NKL", "nkl."),
            SnlSource("NBL", "nbl."),
            SnlSource("SML", "sml."))
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    class Article {
        @JsonProperty("article_url")
        lateinit var articleUrl: String
        lateinit var headword: String
        lateinit var snippet: String
        @JsonProperty("taxonomy_title")
        lateinit var taxonomyTitle: String
    }
}